programs/pluto/cert_rotation.c

```diff
@@ -0,0 +1,50 @@
+#include <nss.h>
+
+#include "cert_rotation.h"
+#include "connections.h"
+#include "constants.h"
+#include "defs.h"
+#include "keys.h"
+#include "log.h"
+#include "nss_cert_load.h"
+#include "whack.h"
+
+void rotate_cert(const struct fd *whackfd, const struct whack_message *wm)
+{
+	struct connection *conn;
+	struct end *dst;
+	const char *nickname;
+	enum whack_pubkey_type pubkey_type;
+	CERTCertificate *old_nss_cert;
+
+	conn = conn_by_name(wm->name, false);
+	if (conn == NULL)  {
+		whack_comment(whackfd, "No connection named '%s' found",
+			wm->name);
+		return;
+	}
+
+	dst = &conn->spd.this;
+	nickname = cert_nickname(&dst->cert);
+	if (nickname == NULL) {
+		whack_comment(whackfd, "No cert nickname found");
+		return;
+	}
+
+	// temporarily back up cert to release or restore it later
+	old_nss_cert = dst->cert.u.nss_cert;
```

Comment:
This is relying on load_end_cert_and_preload_secret() being hardwired to:
	dst_end->cert.ty = CERT_NONE;
	dst_end->cert.u.nss_cert = NULL;
I'd instead save the entire end->cert and clear it; and then restore the contents if things fail.
I'd make that explicit here, and remember to restore everything.  And remember to restore

Reply:
good point..  I will fix it.